torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -594,9 +602,16 @@ void ProcessGroupNCCL::ncclCommWatchdog() {
     LOG(INFO) << "[Rank " << rank_
               << "] NCCL watchdog thread terminated normally";
   } catch (std::exception& e) {
+    auto err = e.what();
     LOG(INFO) << "[Rank " << rank_
               << "] NCCL watchdog thread terminated with exception: "
-              << e.what();
+              << err;
+    // Errors containing handleNCCLGuardLogStr indicate that the
+    // process is unhealthy and needs to be taken down.
+    if (strstr(err, handleNCCLGuardLogStr) != nullptr) {
```

Comment:
This is a bit hacky, can we just use `std::exit` on line 592 instead after logging the reason why we are taking the process down?

Reply:
We could do this, but won't it introduce inconsistency between the 2 watchdogs in how they bring down the process for nccl_async_error_handling? I am not sure if it is preferable to log + std::exit or crash with an exception, but we are doing the latter currently.

Reply:
If we want to do this via an exception, let us define a custom exception type for this and throw that. Then here we can have multiple catch statements, one for the custom exception type in which case we rethrow and one for `std::exception` where we perform the rest of the handling.